mitodl/mit-learn
# Mechanism to sync server prefetch with client API calls

### What are the relevant tickets?





Closes https://github.com/mitodl/hq/issues/5918

### Description (What does it do?)


To keep server and client API calls in sync, we need a solution that warns us if API calls are made from the browser to fetch content that could have been pre-rendered on the server, or indeed we're making unnecessary API calls on the server to fetch content that is not in the page. In addition to the warnings, it would ideally offer code reuse between server and client and help with the development workflow, or be minimally intrusive to it.

This PR issues console warnings for API calls made during the initial page render that have not been prefetched on the server and populated into the React Query cache.

- Waits for the document to be loaded and flags it on the query cache.
- Wraps our `useQuery()` hooks with a function that checks for cached data at a given key.
- Log a console warning if the document has not loaded and a key requested by a client component has nothing cached.
- Triggers a check on document load for any queries prefetched but not requested in components during the initial load.


Prefetches API data on the server for: 
- some of the homepage (excludes resources, commented in code)
- the department listing page



### How can this be tested?


- Do not prefetch a query that is needed during initial render, e.g. in frontends/main/src/app/departments/page.tsx, remove `channelsKeyFactory.countsByType("department")` from the prefetch array.

  In the consoles expect to see warning:
```
['channel', 'countsByType', 'department'] Content was not prefetched on the server for query key needed during initial render
```

- Prefetch a query on the server that is not needed for initial render, e.g. in frontends/main/src/app/departments/page.tsx add `channelsKeyFactory.countsByType("random")` to the prefetch array.

  In the consoles expect to see warning:
```
["channel","countsByType","random"] Content was prefetched on the server for query key not needed during initial render 
```

### Additional Context



For discussion, comments below.
